openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -825,6 +849,58 @@ def to_dict(self, output_units=None, return_cosmetic_attributes=False):
         smirnoff_data.update(unitless_header_attribute_dict)
         smirnoff_data.update(output_units)
         return smirnoff_data
+
+    # -------------------------------
+    # Utilities for children classes.
+    # -------------------------------
+
+    @classmethod
+    def _check_all_valence_terms_assigned(cls, assigned_terms, valence_terms):
+        """Check that all valence terms have been assigned and print a user-friendly error message.
+
+        Parameters
+        ----------
+        assigned_terms : ValenceDict
+            Atom index tuples defining added valence terms.
+        valence_terms : Iterable[TopologyAtom] or Iterable[Iterable[TopologyAtom]]
+            Atom or atom tuples defining topological valence terms.
+
+        """
+        # Convert the valence term to a valence dictionary to make sure
+        # the order of atom indices doesn't matter for comparison.
+        valence_terms_dict = assigned_terms.__class__()
+        for atoms in valence_terms:
+            try:
+                # valence_terms is a list of TopologyAtom tuples.
+                atom_indices = (a.topology_particle_index for a in atoms)
+            except TypeError:
+                # valence_terms is a list of TopologyAtom.
+                atom_indices = (atoms.topology_particle_index,)
+            valence_terms_dict[atom_indices] = atoms
+
+        # Check that both valence dictionaries have the same keys (i.e. terms).
+        assigned_terms_set = set(assigned_terms.keys())
+        valence_terms_set = set(valence_terms_dict.keys())
+        unassigned_terms = valence_terms_set.difference(assigned_terms_set)
+        not_found_terms = assigned_terms_set.difference(valence_terms_set)
+
+        # Raise an error if there are unassigned terms.
+        err_msg = ""
+
+        if len(unassigned_terms) > 0:
+            err_msg += ("{parameter_handler} was not able to find parameters for the following valence terms:\n"
+                        "- {unassigned_str}").format(parameter_handler=cls.__name__,
+                                                     unassigned_str='\n- '.join(unassigned_terms))
+        if len(not_found_terms) > 0:
+            if err_msg != "":
+                err_msg += '\n'
+            err_msg += ("{parameter_handler} assigned terms that were not found in the topology:\n"
+                        "- {not_found_str}").format(parameter_handler=cls.__name__,
+                                                    not_found_str='\n- '.join(not_found_terms))
+        if err_msg != "":
+            raise RuntimeError(err_msg)
```

Comment:
I don't think we want to raise `RuntimeError`, since these are not going to be informative about what was not matched. Instead, why don't we define our own exception hierarchy (e.g. `UnassignedParameterException > UnassignedValenceParameterException > UnassignedBondParameterException`) so that applications or tests could catch these exceptions and proceed with execution if intended. (For example, someone might be mixing a force field and NN potential outside of SMIRNOFF.)

Also, let's be sure the docstrings document exceptions we intentionally throw under different circumstances.

Reply:
Will do, thanks!